### PR TITLE
[quantizer] add fc-1d rewrite

### PR DESCRIPTION
**Pull request by zk1998**

### CODE REVIEW

File: tinynn/graph/quantization/quantizer.py [hunk 1]
```diff
@@ -91,6 +95,9 @@
 # Constant func names
 creation_func_names = []
 
+# Processed Rewrite to fuse rules
```
Comment (peterjc123) [inline tinynn/graph/quantization/quantizer.py:98]: Why was capitalized character used here?

File: tinynn/graph/quantization/quantizer.py [hunk 2]
```diff
@@ -327,6 +334,7 @@ def _find_quantized_prelu_nodes(node: TraceNode, custom_node):
         graph.filter_forward_nodes(is_fusable, custom_data, reverse=True)
         quant_list = custom_data[0]
         log.info(f'found nodes to fuse: {quant_list}')
+        print(f'found nodes to fuse: {quant_list}')
```
Comment (peterjc123) [inline tinynn/graph/quantization/quantizer.py:337]: Please don't use `print` here.

File: tinynn/graph/quantization/quantizer.py [hunk 3]
```diff
@@ -1306,6 +1314,85 @@ def _avgpool_kernel_size_and_stride(kernel_size, stride=None, *args, **kwargs):
...
+        for idx, names in enumerate(reversed(rewrite_fuse_names_list)):  # names should be ['fc', 'bn']
+            # case fc-bn1d
+            if len(names) == 2:
+                node_fc = graph.nodes_map[names[0]]
+                node_bn1d = graph.nodes_map[names[1]]
+                mod_fc = node_fc.module
+                mod_bn = node_bn1d.module
+                if type(mod_fc) == nn.Linear and type(mod_bn) == nn.BatchNorm1d:
+                    if len(node_fc.prev_tensors[0].shape) != 2:
+                        log.info('the fc-bn\'s input dimension != 2')
```
Comment (peterjc123) [inline tinynn/graph/quantization/quantizer.py:1338]: Consider using the`DEBUG` level here. It's a confusing message for the users.

File: tests/quantizer_test.py [hunk 4]
```diff
@@ -784,6 +784,23 @@ def forward(self, x):
 
         check_quantize_rewrite(model, inputs)
 
+    def test_fc_bn_fuse(self):
```
Comment (peterjc123) [inline tests/quantizer_test.py:787]: Please test Linear with `bias=False` as well.

File: tinynn/graph/quantization/quantizer.py [hunk 5]
```diff
@@ -1306,6 +1314,85 @@ def _avgpool_kernel_size_and_stride(kernel_size, stride=None, *args, **kwargs):
...
+            current_rules=load_processed_rewrite_to_fuse_rules(),
+            check_node_quantized=False,
+            graph=graph,
+        )
+        custom_data = ([], set())
+        graph.filter_forward_nodes(is_rewrite_to_fuse, custom_data, reverse=True)
+        rewrite_fuse_names_list = custom_data[0]
+        log.debug(f'found names_list that need to rewrite for fusing: {rewrite_fuse_names_list}')
+
+        for idx, names in enumerate(reversed(rewrite_fuse_names_list)):  # names should be ['fc', 'bn']
```
Comment (peterjc123) [inline tinynn/graph/quantization/quantizer.py:1329]: Please remove the comment in this line.

File: tinynn/graph/quantization/quantizer.py [hunk 6]
```diff
@@ -1306,6 +1314,85 @@ def _avgpool_kernel_size_and_stride(kernel_size, stride=None, *args, **kwargs):
...
+            # case fc-bn1d
+            if len(names) == 2:
```
Comment (peterjc123) [inline tinynn/graph/quantization/quantizer.py:1331]: If you add this check, please remember to handle the `else` codepath.

File: tinynn/graph/quantization/quantizer.py [hunk 7]
```diff
@@ -1306,6 +1314,85 @@ def _avgpool_kernel_size_and_stride(kernel_size, stride=None, *args, **kwargs):
...
+                node_fc = graph.nodes_map[names[0]]
+                node_bn1d = graph.nodes_map[names[1]]
+                mod_fc = node_fc.module
+                mod_bn = node_bn1d.module
+                if type(mod_fc) == nn.Linear and type(mod_bn) == nn.BatchNorm1d:
```
Comment (peterjc123) [inline tinynn/graph/quantization/quantizer.py:1336]: If you add this check, please remember to handle the `else` codepath.

File: tinynn/graph/quantization/quantizer.py [hunk 8]
```diff
@@ -2008,3 +2095,20 @@ def load_processed_extra_qat_rules():
             base_rule_pair[0] = True
         processed_extra_qat_rules.update(rule_dict)
     return processed_extra_qat_rules
+
+
+def load_processed_rewrite_to_fuse_rules():
```
Comment (peterjc123) [inline tinynn/graph/quantization/quantizer.py:2100]: Consider refactoring the functions. As can be seen that Line 2071-2079, 2087- 2097 and 2104-2114 are almost the same.

**Review by peterjc123 — CHANGES_REQUESTED** [review]

**Comment by zk1998** [comment]
您好，我已经收到您的邮件，我会尽快查看并在第一时间回复您。

**Review by peterjc123 — APPROVED** [review]

**Comment by zk1998** [comment]
您好，我已经收到您的邮件，我会尽快查看并在第一时间回复您。

**Pull request merged**